cocotb/share/makefiles/simulators/Makefile.aldec

```diff
@@ -105,9 +105,6 @@ $(SIM_BUILD)/runsim.tcl : $(VERILOG_SOURCES) $(VHDL_SOURCES) $(SIM_BUILD)
 	@echo "	puts [read [open sim.log r]]" >> $@
 	@echo "	quit -code 1" >> $@
 	@echo "}" >> $@
-	@echo "@if [string length [array get env LICENSE_QUEUE]] {" >> $@
```

Comment:
Can't we keep this and just remove the extra `@`?

Reply:
I tried and for me this is also not working, leads to 
```
sim_build/runsim.tcl(5) error: parse error
if [string length<--
/c/temp/cocotb/cocotb/share/makefiles/simulators/Makefile.aldec:145: recipe for target 'results.xml' failed
make[1]: *** [results.xml] Error 1
```

Reply:
I think this perhaps should be:
```
@echo "if {[string length [array get env LICENSE_QUEUE]]} {" >> $@
```

Reply:
Or maybe even just
```
@echo "if {[string length $$::env(LICENSE_QUEUE)]} {" >> $@
```

Reply:
Both not working for me:
```
sim_build/runsim.tcl(5) error: parse error
if {[string length<--
```

Reply:
Maybe
```
ifeq ($(LICENSE_QUEUE),)
        @echo "if [string length [array get env LICENSE_QUEUE]] {" >> $@
	@echo " set LICENSE_QUEUE $$::env(LICENSE_QUEUE)" >> $@
	@echo "}" >> $@
endif
```

Reply:
Something really weird seems to be happening with your TCL intepreter @pstrueb. 

Can you add `puts $tcl_version` to this file and see what it prints?

Reply:
Or better:

```
ifneq ($(LICENSE_QUEUE),)
	@echo "set LICENSE_QUEUE $$::env(LICENSE_QUEUE)" >> $@
endif
```

so effect the same as before but will not make issues.

Reply:
@themperek: I don't think that works if you change the variable after forgetting to set it for the first time. To elaborate, with your proposed change:
```
$ make sim
error: no license
$ LICENSE_QUEUE=... make sim
# does not rebuild the tcl file
error: no license
``
I'd rather debug why legal tcl code isn't working that put in a fragile workaround.

Reply:
>after forgetting to set it
I would think this is a very minor issue.

Reply:
> Something really weird seems to be happening with your TCL intepreter @pstrueb.
> 
> Can you add `puts $tcl_version` to this file and see what it prints?

Hm thats not working for me. It rather gives `cl_version` or only whitespace for $(tcl_version).



> Maybe
> 
> ```
> ifneq ($(LICENSE_QUEUE),)
>         @echo "if [string length [array get env LICENSE_QUEUE]] {" >> $@
> 	@echo " set LICENSE_QUEUE $$::env(LICENSE_QUEUE)" >> $@
> 	@echo "}" >> $@
> endif
> ```
> 
> As some workaround.

This leads to
```
ifneq (,)
bash: -c: line 0: syntax error near unexpected token `,'
bash: -c: line 0: `ifneq (,)'
/c/temp/cocotb/cocotb/share/makefiles/simulators/Makefile.aldec:104: recipe for target 'sim_build/runsim.tcl' failed
```
if I implement it like this:
```
$(SIM_BUILD)/runsim.tcl : $(VERILOG_SOURCES) $(VHDL_SOURCES) $(SIM_BUILD)
	@echo "onerror {" > $@
	@echo "	puts [read [open sim.log r]]" >> $@
	@echo "	quit -code 1" >> $@
	@echo "}" >> $@
	ifneq ($(LICENSE_QUEUE),)
		@echo "if [string length [array get env LICENSE_QUEUE]] {" >> $@
		@echo " set LICENSE_QUEUE $$::env(LICENSE_QUEUE)" >> $@
		@echo "}" >> $@
	endif	
	@echo "alib $(RTL_LIBRARY)" >> $@
	@echo "set worklib $(RTL_LIBRARY)" >> $@
```

Reply:
> Something really weird seems to be happening with your TCL intepreter @pstrueb.
> 
> Can you add `puts $tcl_version` to this file and see what it prints?

To be clear, I'm suggesting
```suggestion
    @echo "puts $$tcl_version" >> $@
```
or directly editing the `runsim.tcl` file manually to contain `puts $tcl_version`

Reply:
> > Something really weird seems to be happening with your TCL intepreter @pstrueb.
> > Can you add `puts $tcl_version` to this file and see what it prints?
> 
> To be clear, I'm suggesting
> 
> or directly editing the `runsim.tcl` file manually to contain `puts $tcl_version`
```
$(SIM_BUILD)/runsim.tcl : $(VERILOG_SOURCES) $(VHDL_SOURCES) $(SIM_BUILD)
	@echo "onerror {" > $@
	@echo "	puts [read [open sim.log r]]" >> $@
	@echo "	quit -code 1" >> $@
	@echo "}" >> $@
	@echo "puts $$tcl_version" >> $@
	@echo "alib $(RTL_LIBRARY)" >> $@
	@echo "set worklib $(RTL_LIBRARY)" >> $@
```

leads to a runsim.tcl with
```
onerror {
	puts [read [open sim.log r]]
	quit -code 1
}
puts 
alib work
set worklib work
acom  -dbg C:/temp/cocotb/examples/adder/hdl/adder.vhdl
asim   +access +w -interceptcoutput -O2 -dbg -loadvhpi "C:/temp/cocotb/cocotb/libs/libcocotbvhpi_aldec:vhpi_startup_routines_bootstrap" adder 
run -all
endsim
```

Not sure if this is getting us anywhere...

Reply:
Can you just edit the file directly? Make shouldn't rebuild and overwrite it

Reply:
Alternatively, this might work:
```suggestion
	@echo 'puts $$tcl_version' >> $@
```

Reply:
> Can you just edit the file directly? Make shouldn't rebuild and overwrite it

so If I edit the file directly to

```
onerror {
	puts [read [open sim.log r]]
	quit -code 1
}
puts $tcl_version
alib work
set worklib work
acom  -dbg C:/temp/cocotb/examples/adder/hdl/adder.vhdl
asim   +access +w -interceptcoutput -O2 -dbg -loadvhpi "C:/temp/cocotb/cocotb/libs/libcocotbvhpi_aldec:vhpi_startup_routines_bootstrap" adder 
run -all
endsim
```
it leads to 
```
(py37_32) PS C:\temp\cocotb\examples\adder\tests> make SIM=aldec TOPLEVEL_LANG=vhdl
/usr/bin/make results.xml
make[1]: Entering directory '/c/temp/cocotb/examples/adder/tests'
set -o pipefail; GPI_EXTRA= TOPLEVEL_LANG=vhdl \
 MODULE=test_adder TESTCASE= TOPLEVEL=adder COCOTB_SIM=1 /c/Aldec/Active-HDL9.3/BIN/vsimsa  -do sim_build/runsim.tcl | tee sim_build/sim.log

VHDL/Verilog/EDIF/SystemC Simulator build 9.3.2744.sp1.09.4995

(c) 1997-2014 Aldec, Inc. All rights reserved.

License Number 0

VSIMSA: Configuration files: `C:\Aldec\Active-HDL9.3\vlib\library.cfg', `C:\Aldec\Active-HDL9.3\BIN\vsimsa.cfg'
 Welcome to VSIMSA!
 This message was printed from `startup.do' macro file.

puts $tcl_version
sim_build/runsim.tcl(5) Error: Wrong variable name `$tcl_version'.
```

(py37_32) PS C:\temp\cocotb\examples\adder\tests> make SIM=aldec TOPLEVEL_LANG=vhdl
/usr/bin/make results.xml
make[1]: Entering directory '/c/temp/cocotb/examples/adder/tests'
set -o pipefail; GPI_EXTRA= TOPLEVEL_LANG=vhdl \
 MODULE=test_adder TESTCASE= TOPLEVEL=adder COCOTB_SIM=1 /c/Aldec/Active-HDL9.3/BIN/vsimsa  -do sim_build/runsim.tcl | tee sim_build/sim.log

VHDL/Verilog/EDIF/SystemC Simulator build 9.3.2744.sp1.09.4995

(c) 1997-2014 Aldec, Inc. All rights reserved.

License Number 0

VSIMSA: Configuration files: `C:\Aldec\Active-HDL9.3\vlib\library.cfg', `C:\Aldec\Active-HDL9.3\BIN\vsimsa.cfg'
 Welcome to VSIMSA!
 This message was printed from `startup.do' macro file.

puts $tcl_version
sim_build/runsim.tcl(5) Error: Wrong variable name `$tcl_version'.
        puts [read [open sim.log r]]
sim_build/runsim.tcl(2) Error: Unknown command `puts'
/c/temp/cocotb/cocotb/share/makefiles/simulators/Makefile.aldec:142: recipe for target 'results.xml' failed
make[1]: *** [results.xml] Error 1
make[1]: Leaving directory '/c/temp/cocotb/examples/adder/tests'
c:\temp\cocotb\cocotb\share\makefiles/Makefile.sim:153: recipe for target 'sim' failed
make: *** [sim] Error 2

Reply:
> > after forgetting to set it
> 
> I would think this is a very minor issue.

Same is the case for SCRIPT_FILE, CFG_TOPLEVEL, WAVES, GUI and COVERAGE

Reply:
Thanks for testing. Something seems very weird with the aldec shell if it doesn't even have puts.

`info patchlevel` might give something useful, but without a puts I don't see how to even get the information.

If your aldec is behaving so differently to @themperek's, I'm thinking we maybe should do things conditional on which aldec is being used.

Reply:
@eric-wieser Those 3 lines are straight copy from Riviera documentation (I just checked). It just defines the time it will wait for a license before exit.  
PS. `@` is not a mistake this is like this in the documentation.

I this case I think this should be good enough:
```
ifneq ($(LICENSE_QUEUE),)
    @echo "set LICENSE_QUEUE $$::env(LICENSE_QUEUE)" >> $@
endif
```

@pstrueb  Mind the `tab`s.

Reply:
Thanks for the debugging, I think the conclusion is that aldec active-hdl does not have a real TCL intepreter, while aldec riviera does.

Reply:
The problem is that Active-HDL has several modes for it's interpreter. the `scripterconf` command is used to display and edit the mode. The `-do` argument to `VSimSA.exe` runs the script in **DO** mode, which is a language based on TCL but not quite TCL.

There are further complications with Active-HDL 10.1 regarding the use of the `-loadvhpi` argument to `asim` command, as well as a cocotb-breaking bug with 10.4 through 10.5a (see #1494).